src/surface_finder.cpp

```diff
@@ -0,0 +1,316 @@
+  // 3. Store the plane and rank it by number of points the plane covers
+
+  ROS_INFO("Start exploring surfaces in %ld indices of %s",
+           cloud_indices_->indices.size(),
+           cloud_indices_->header.frame_id.c_str());
+
+  size_t num_surface = 0;
+  size_t max_inlier_count = std::numeric_limits<size_t>::min();
+  std::srand(unsigned(std::time(0)));
+
+  std::map<size_t,
+           std::pair<pcl::ModelCoefficients::Ptr, pcl::PointIndices::Ptr>,
+           std::greater<size_t> >
+      ranking;
+
+  SurfaceHistoryRecorder recorder;
+
+  // Timer start point
+  clock_t start = std::clock();
+
+  // Sample min_iteration_ of horizontal surfaces
+  while (num_surface < min_iteration_ || ranking.size() < min_surface_amount) {
+    pcl::ModelCoefficients::Ptr coeff(new pcl::ModelCoefficients);
+    coeff->values.resize(4);
+    pcl::PointIndices::Ptr indices(new pcl::PointIndices);
+
+    // Sample 1 point randomly to establish a plane, by assuming it's horizontal
+    size_t rand_index = std::rand() % cloud_indices_->indices.size();
+    const PointC& pt = cloud_->points[cloud_indices_->indices[rand_index]];
+
+    // Count points within given distance to the plane
+    std::vector<int> inlier_indices;
+    filterIndices(max_point_distance_, sorted_indices_, pt.z, &inlier_indices);
+
+    // Establish coefficients
+    coeff->values[0] = 0;
+    coeff->values[1] = 0;
+    coeff->values[2] = 1;
+    coeff->values[3] = -1 * pt.z;
+    indices->indices = inlier_indices;
+
+    // Check surface point threshold
+    if (indices->indices.size() > surface_point_threshold_) {
+      // Update plane choices
+      bool qualify = true;
+      bool pre_exist = false;
+      size_t old_indices_size;
+      for (std::map<
+               size_t,
+               std::pair<pcl::ModelCoefficients::Ptr, pcl::PointIndices::Ptr>,
+               std::greater<size_t> >::iterator it = ranking.begin();
+           it != ranking.end(); it++) {
+        pcl::ModelCoefficients& old_coeff = *(it->second.first);
+
+        old_indices_size = it->first;
+
+        // If newly found surface is better, replace
+        if (isSimilar(max_point_distance_, old_coeff, *coeff)) {
+          if (old_indices_size < indices->indices.size()) {
+            ranking.erase(it);
+            pre_exist = true;
+            break;
+          } else {
+            qualify = false;
+          }
+        }
+      }
+      if (qualify) {
+        std::pair<pcl::ModelCoefficients::Ptr, pcl::PointIndices::Ptr> pr(
+            coeff, indices);
+        ranking[indices->indices.size()] = pr;
+        if (pre_exist) {
+          recorder.Update(old_indices_size, indices->indices.size(), cloud_,
+                          indices, num_surface);
+        } else {
+          recorder.Record(indices->indices.size(), cloud_, indices,
+                          num_surface);
+        }
+      }
+    }
+
+    num_surface++;
+  }
+
+  //  Report surfaces
+  if (ranking.size() > 0) {
```

Comment:
As I mentioned elsewhere, I think that the history should be optional. You could just add:

bool debug=false

At the beginning of this method. Then, if debug is true, create and maintain the history. We want the algorithm to be as fast as possible, and it looks like the history class is copying point cloud data at least a few times.

Reply:
Completed